src/Pate/Equivalence/MemoryDomain.hs

```diff
@@ -0,0 +1,224 @@
+traverseWithCellPar ::
+  forall sym arch m future.
+  PP.IsFuture m future =>
+  W4.IsExprBuilder sym =>
+  MemoryDomain sym arch ->
+  (forall w. 1 <= w => PMC.MemCell sym arch w -> W4.Pred sym -> m (future (W4.Pred sym))) ->
+  m (future (MemoryDomain sym arch))
+traverseWithCellPar memDom f = do
+  let
+    f' :: Some (PMC.MemCell sym arch) -> W4.Pred sym -> m (future (W4.Pred sym))
+    f' (Some cell@(PMC.MemCell{})) p = f cell p
+
+  future_preds <- M.traverseWithKey f' (memDomainPred memDom)
+  PP.present $ do
+    preds <- traverse PP.joinFuture future_preds
+    return $ MemoryDomain (PMC.dropTrivialCells preds) (memDomainPolarity memDom)
+
+      
+traverseWithCell ::
+  forall sym arch m.
+  Monad m =>
+  W4.IsExprBuilder sym =>
+  MemoryDomain sym arch ->
+  (forall w. 1 <= w => PMC.MemCell sym arch w -> W4.Pred sym -> m (W4.Pred sym)) -> m (MemoryDomain sym arch)
+traverseWithCell memDom f = join $ traverseWithCellPar memDom (\cell p -> return $ f cell p)
+
+
+toList ::
+  MemoryDomain sym arch ->
+  [(Some (PMC.MemCell sym arch), W4.Pred sym)]
+toList memDom = M.toList (memDomainPred memDom)
+
+toAscList ::
+  MemoryDomain sym arch ->
+  [(Some (PMC.MemCell sym arch), W4.Pred sym)]
+toAscList memDom = M.toAscList (memDomainPred memDom)
+
+cells ::
+  OrdF (W4.SymExpr sym) =>
+  MemoryDomain sym arch ->
+  Set (Some (PMC.MemCell sym arch))
+cells memDom = S.fromList $ map fst (toList memDom)
+
+-- | Build a 'MemoryDomain' from a list of 'PMC.MemCell' entries, with a corresponding
+-- predicate indicating whether or not they are in or out of the domain (subject to the
+-- given polarity).
+-- Duplicate entries are allowed, where the resulting predicate will be the disjunction
+-- of the overlapping predicates.
+fromList ::
+  W4.IsExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  sym ->
+  [(Some (PMC.MemCell sym arch), W4.Pred sym)] ->
+  W4.Pred sym ->
+  IO (MemoryDomain sym arch)
+fromList sym l pol = do
+  -- NOTE: We can't just use Data.Map.fromList here because it will discard duplicate
+  -- entries
+  let maps = map (\(cell, p) -> M.singleton cell p) l
+  merged <- foldM (PMC.mergeMemCellPred sym) M.empty maps
+  return $ MemoryDomain merged pol
+
+-- | Similar to 'fromList' but assumes the cells are in order and disjoint
+fromAscList ::
+  W4.IsExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  [(Some (PMC.MemCell sym arch), W4.Pred sym)] ->
+  W4.Pred sym ->
+  MemoryDomain sym arch
+fromAscList l pol = MemoryDomain (M.fromAscList l) pol
+
+mux ::
+  W4.IsExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  sym ->
+  W4.Pred sym ->
+  MemoryDomain sym arch ->
+  MemoryDomain sym arch ->
+  IO (MemoryDomain sym arch)
+mux sym p predT predF = case W4.asConstantPred p of
+  Just True -> return predT
+  Just False -> return predF
+  _ -> do
+    pol <- W4.baseTypeIte sym p (memDomainPolarity predT) (memDomainPolarity predF)
+    locs <- PMC.muxMemCellPred sym p (memDomainPred predT) (memDomainPred predF)
```

Comment:
Actually I realized that this invariant doesn't hold, since our "external" domains flip the polarity of the stack to demand exact equivalence. I'll write up a ticket about it and update the comment, but no we can't just assert that the polarity is always concrete.